packages/services/src/state_listener.rs

```diff
@@ -0,0 +1,294 @@
+use async_trait::async_trait;
+use ports::{
+    storage::Storage,
+    types::{SubmissionTx, TransactionReceipt, TransactionState},
+};
+
+use super::Runner;
+
+pub struct StateListener<L1, Db> {
+    l1_adapter: L1,
+    storage: Db,
+    num_blocks_to_finalize: u64,
+}
+
+impl<L1, Db> StateListener<L1, Db> {
+    pub fn new(l1_adapter: L1, storage: Db, num_blocks_to_finalize: u64) -> Self {
+        Self {
+            l1_adapter,
+            storage,
+            num_blocks_to_finalize,
+        }
+    }
+}
+
+impl<L1, Db> StateListener<L1, Db>
+where
+    L1: ports::l1::Api,
+    Db: Storage,
+{
+    async fn check_pending_txs(&mut self, pending_txs: Vec<SubmissionTx>) -> crate::Result<()> {
```

Comment:
Consider the following:
1. there is 1 pending tx
2. we ask for its receipt get back Some(receipt)
3. we determine that it has no `status` (you don't know if it succeeded or not)
4. we determine that enough blocks have been mined afterwards
5. we mark a transaction of unknown status as `Finalized`.

Reply:
solved in the previous comment